snap/validate.go

```diff
@@ -198,9 +198,19 @@ func validateSocketAddrPath(socket *SocketInfo, fieldName string, path string) e
 		return fmt.Errorf("invalid %q: %q should be written as %q", fieldName, path, clean)
 	}
 
-	if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/") || strings.HasPrefix(path, "$XDG_RUNTIME_DIR/")) {
-		return fmt.Errorf(
-			"invalid %q: must have a prefix of $SNAP_DATA, $SNAP_COMMON or $XDG_RUNTIME_DIR", fieldName)
+	switch socket.App.DaemonScope {
+	case SystemDaemon:
+		if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/") || strings.HasPrefix(path, "$XDG_RUNTIME_DIR/")) {
```

Comment:
Curiously, I think this is wrong. System daemons cannot refer to $XDG_RUNTIME_DIR, as that is only provided to sessions. On my Focal system system services no longer run in sessions so they do not have XDG_RUNTIME_DIR.

I would be happy to have a follow up that deals with this but please add a // TODO comment if you are going to push any more patches into this branch.

Reply:
I don't think it makes sense either: given that systemd mounts a tmpfs over a user's $XDG_RUNTIME_DIR, it isn't clear the system instance of systemd can reliably provide sockets in that location (even ignoring the question of which user's XDG_RUNTIME_DIR it should refer to).

FWIW, it looks like this was introduced in PR #6327.  I left it unchanged since there are probably some snaps making use of it right now (maybe something from the Mir team?).

Reply:
We should look, after this PR, because I suspect it expands to something silly (unless we expand it ignoring the environment).